File: OJ practices/LeetCode题目集/264. Ugly Number II(medium).cpp
/*
 * @Author: SourDumplings
 * @Date: 2021-01-17 09:20:57
 * @Link: https://github.com/SourDumplings/
 * @Email: changzheng300@foxmail.com
 * @Description: https://leetcode-cn.com/problems/ugly-number-ii/
 */

class Solution
{
public:
    int nthUglyNumber(int n)
    {
        // 这三个指针可理解为指向乘以对应因子得到下一个数的位置
        int i2 = 0, i3 = 0, i5 = 0;
        int nums[1690];
        nums[0] = 1;
        for (int i = 1; i < 1690; ++i)
        {
            int ugly = min(min(nums[i2] * 2, nums[i3] * 3), nums[i5] * 5);
            nums[i] = ugly;
            if (ugly == nums[i2] * 2)
            {
                ++i2;
            }
            if (ugly == nums[i3] * 3)
            {
                ++i3;
            }
            if (ugly == nums[i5] * 5)
            {
                ++i5;
            }
        }
        return nums[n - 1];
    }
};

File: OJ practices/LeetCode题目集/264. Ugly Number II(medium).py
'''
Author: SourDumplings
Date: 2021-03-20 11:20:04
Link: https://github.com/SourDumplings/
Email: changzheng300@foxmail.com
Description: https://leetcode-cn.com/problems/ugly-number-ii/
'''


class Solution:
    def nthUglyNumber(self, n: int) -> int:
        uglyNums = [1]
        primes = [2, 3, 5]
        indexes = [0] * 3  # 记录与2，3，5乘的下一个 uglyNum 的索引
        while len(uglyNums) < n:
            nextUglyNum = min(
                [primes[i] * uglyNums[indexes[i]] for i in range(3)])
            uglyNums.append(nextUglyNum)
            for i in range(3):
                if primes[i] * uglyNums[indexes[i]] == nextUglyNum:
                    indexes[i] += 1
        return uglyNums[-1]